// ts-to-jsdoc/bin/ts-to-jsdoc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
const arg_1 = __importDefault(require("arg"));
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const index_1 = __importDefault(require("../index"));
const { "--out": out, "--ignore": ignore, "--force": force, "--help": help, _, } = (0, arg_1.default)({
    "--out": String,
    "-o": "--out",
    "--output": "--out",
    "--ignore": [String],
    "-i": "--ignore",
    "--force": Boolean,
    "-f": "--force",
    "--help": Boolean,
    "-h": "--help",
});
const args = {
    out, ignore, force, help, _,
};
const helpMessage = `
Usage:
  ts-to-jsdoc [options] <path>...

Options:
  -h --help          Shows this.
  -o --out --output  Directory to output transpiled JavaScript. [default: source path]
  -i --ignore        File or directory paths to ignore when transpiling.
  -f --force         Overwrite existing output files.
`;
if (args.help || Object.keys(args).every((arg) => { var _a; return !((_a = args[arg]) === null || _a === void 0 ? void 0 : _a.length); })) {
    console.log(helpMessage);
    process.exit(0);
}
if (args.out) {
    args.out = makePathAbsolute(args.out);
    if (fs_1.default.existsSync(args.out)) {
        if (!args.force) {
            console.error(error(`Output directory exists: ${args.out}`));
            process.exit(1);
        }
    }
    else {
        fs_1.default.mkdirSync(args.out);
    }
}
args.ignore = ((_a = args.ignore) === null || _a === void 0 ? void 0 : _a.length) ? normalizePaths(args.ignore) : [];
const paths = replaceDirectoriesWithFiles([...new Set(normalizePaths(args._))])
    .filter((filepath) => path_1.default.extname(filepath) === ".ts" && !filepath.endsWith(".d.ts"))
    .filter((filepath) => !args.ignore.some((ignoredPath) => filepath === ignoredPath || pathIsInside(filepath, ignoredPath)));
for (const filepath of paths) {
    const outPath = path_1.default.join((_b = args.out) !== null && _b !== void 0 ? _b : path_1.default.dirname(filepath), `${path_1.default.basename(filepath, ".ts")}.js`);
    if (fs_1.default.existsSync(outPath) && !args.force) {
        console.warn(warning(`Cannot write to ${outPath}; file already exists.`));
        continue;
    }
    const code = fs_1.default.readFileSync(filepath, "utf8");
    const transpiled = (0, index_1.default)(code, filepath, {}, true);
    fs_1.default.writeFileSync(outPath, transpiled);
}
function warning(message) {
    return `\u001B[93m[WARN]\u001B[39m ${message}`;
}
function error(message) {
    return `\u001B[91m[ERROR]\u001B[39m ${message}`;
}
function makePathAbsolute(filepath) {
    return path_1.default.isAbsolute(filepath)
        ? filepath
        : path_1.default.resolve(process.cwd(), filepath);
}
/**
 * Makes paths absolute, filtering those that exist
 * @return An array containing absolute paths that do exist
 * @param pathsToNormalize
 */
function normalizePaths(pathsToNormalize) {
    return pathsToNormalize
        .map(makePathAbsolute)
        .filter((filepath) => {
        if (!fs_1.default.existsSync(filepath)) {
            console.warn(warning(`File or directory ${filepath} does not exist.`));
            return false;
        }
        return true;
    });
}
/**
 * Given an array of paths, recursively removes all
 * directories and appends all files within said directories to the array
 * @param pathList An array of paths
 * @return An array containing only files, replacing directories with their contents
 */
function replaceDirectoriesWithFiles(pathList) {
    let pathArray = [...pathList];
    for (const [index, filepath] of pathArray.entries()) {
        if (fs_1.default.existsSync(filepath) && fs_1.default.lstatSync(filepath).isDirectory()) {
            pathArray.splice(index, 1);
            pathArray = pathArray
                .concat(replaceDirectoriesWithFiles(fs_1.default.readdirSync(filepath).map((file) => path_1.default.join(filepath, file))));
        }
    }
    return pathArray;
}
/* eslint-disable */
/**
 * @license WTFPL
 * Copyright © 2013–2016 Domenic Denicola <d@domenic.me>
 */
function pathIsInside(thePath, potentialParent) {
    // For inside-directory checking, we want to allow trailing slashes, so normalize.
    thePath = stripTrailingSep(thePath);
    potentialParent = stripTrailingSep(potentialParent);
    // Node treats only Windows as case-insensitive in its path module; we follow those conventions.
    if (process.platform === "win32") {
        thePath = thePath.toLowerCase();
        potentialParent = potentialParent.toLowerCase();
    }
    return thePath.lastIndexOf(potentialParent, 0) === 0
        && (thePath[potentialParent.length] === path_1.default.sep
            || thePath[potentialParent.length] === undefined);
}
/**
 * @license WTFPL
 * Copyright © 2013–2016 Domenic Denicola <d@domenic.me>
 */
function stripTrailingSep(thePath) {
    if (thePath[thePath.length - 1] === path_1.default.sep) {
        return thePath.slice(0, -1);
    }
    return thePath;
}
